File: .mssql-scripts/check_ballooning_numbering.py
"""Does the sub-number survive a server-side reorder?

The reorder in RPA/API/BalloonPosition.py renumbers a whole page from a
counter. Before this change it gave every item its own integer, which turned
5-1 into a plain 6 and pushed everything after it along by one - the exact
thing a sub-number exists to prevent.

Runs the real recalculate_balloon_position against a synthetic page.

    python .mssql-scripts/check_ballooning_numbering.py
"""

import os
import sys

API = r"C:\Aizera\RPA\API"
sys.path.insert(0, API)
os.chdir(API)

fails = []


def check(name, ok, detail=""):
    print(f"  {'PASS' if ok else 'FAIL'}  {name}{'  ' + detail if detail else ''}")
    if not ok:
        fails.append(name)


def ann(_id, number, sub=None, y=0.0, x=0.0, note=False, section=""):
    return {
        "id": _id, "balloonNumber": number, "subNumber": sub,
        "balloonX": x, "balloonY": y,
        "rect": {"x": x, "y": y, "width": 1.0, "height": 1.0},
        "content": f"dim {_id}", "originalContent": "", "type": "Dimension",
        "status": "Pending", "pageIndex": 0, "auto": True, "isNote": note,
        "section": section, "gridStart": "", "gridEnd": "", "viewId": 0,
    }


print("1. import the processor")
from BalloonPosition import SmartPositioner, GridSorter      # noqa: E402
sp = SmartPositioner()
check("SmartPositioner constructs", True)

print("\n2. reorder keeps children off the integer sequence")
# Three parents top-to-bottom, with two children hanging off the middle one.
page = [
    ann("a", 1, y=10.0),
    ann("b", 2, y=20.0),
    ann("b1", 2, sub=1, y=21.0),
    ann("b2", 2, sub=2, y=22.0),
    ann("c", 3, y=30.0),
]
out = {r["id"]: r for r in sp.recalculate_balloon_position(list(page))}
check("every balloon comes back", len(out) == 5, f"{len(out)} of 5")
got = {k: (v["balloonNumber"], v.get("subNumber")) for k, v in out.items()}
print(f"        {got}")

parents = sorted(v["balloonNumber"] for k, v in out.items()
                 if not v.get("subNumber"))
check("parents get a dense 1..N sequence", parents == [1, 2, 3], str(parents))
check("child b1 keeps its parent's number",
      out["b1"]["balloonNumber"] == out["b"]["balloonNumber"],
      f"{out['b1']['balloonNumber']} vs parent {out['b']['balloonNumber']}")
check("child b2 keeps its parent's number",
      out["b2"]["balloonNumber"] == out["b"]["balloonNumber"])
check("children keep their own sub index",
      (out["b1"]["subNumber"], out["b2"]["subNumber"]) == (1, 2))
check("the balloon after the children is the next integer, not +2",
      out["c"]["balloonNumber"] == out["b"]["balloonNumber"] + 1,
      f"c={out['c']['balloonNumber']}, b={out['b']['balloonNumber']}")

print("\n3. a reorder that actually moves things")
# Same page, but the numbers arrive shuffled relative to position.
shuffled = [
    ann("a", 7, y=10.0),
    ann("b", 3, y=20.0),
    ann("b1", 3, sub=1, y=21.0),
    ann("c", 5, y=30.0),
]
out2 = {r["id"]: r for r in sp.recalculate_balloon_position(list(shuffled))}
print(f"        {[(k, v['balloonNumber'], v.get('subNumber')) for k, v in out2.items()]}")
check("child follows its parent to the parent's NEW number",
      out2["b1"]["balloonNumber"] == out2["b"]["balloonNumber"],
      f"child={out2['b1']['balloonNumber']} parent={out2['b']['balloonNumber']}")
check("parents still dense",
      sorted(v["balloonNumber"] for k, v in out2.items()
             if not v.get("subNumber")) == [1, 2, 3])

print("\n4. an orphaned child keeps its own number")
orphan = [ann("a", 1, y=10.0), ann("x1", 9, sub=1, y=50.0)]
out3 = {r["id"]: r for r in sp.recalculate_balloon_position(list(orphan))}
check("orphan is not silently adopted by another balloon",
      out3["x1"]["balloonNumber"] == 9,
      f"got {out3['x1']['balloonNumber']}, expected the 9 it arrived with")

print("\n5. GridSorter still orders by the sheet's grid")
grid = GridSorter.parse_grid_system("F2", "A1")
check("F2->A1 parses", grid is not None, str(grid))
if grid:
    check("both axes read as descending",
          grid["row_direction"] == -1 and grid["col_direction"] == -1)
    k_placed = GridSorter.get_sort_key(
        {"Section": "D8", "ViewId": 0, "BBoxY1": 5.0, "BBoxX1": 5.0}, grid)
    k_unmatched = GridSorter.get_sort_key(
        {"Section": "UNMATCHED", "ViewId": 0, "BBoxY1": 1.0, "BBoxX1": 1.0}, grid)
    check("a placed cell sorts before an unmatched one",
          k_placed < k_unmatched, f"{k_placed} < {k_unmatched}")

print(f"\n{'ALL PASS' if not fails else str(len(fails)) + ' FAILED: ' + ', '.join(fails)}")
sys.exit(1 if fails else 0)

File: .mssql-scripts/check_ballooning_numbering.mjs
/**
 * Exercises BallooningNumbering.ts against the same cases as the Python side,
 * so the browser's ordering and the server's cannot silently diverge.
 *
 * Compiles the module with the esbuild already in DSRFQ.Web/node_modules -
 * nothing extra to install.
 *
 *     node .mssql-scripts/check_ballooning_numbering.mjs
 */

import { build } from '../DSRFQ.Web/node_modules/esbuild/lib/main.js';
import { pathToFileURL } from 'node:url';
import { mkdtempSync, rmSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { fileURLToPath as __f2p } from 'node:url';
import { dirname as __dn, resolve as __rs } from 'node:path';

// The settings store talks to a Serenity service, which needs a browser. These
// modules only need its values, so the bundle uses the in-memory stub instead.
const stubStore = {
    name: 'stub-settings-store',
    setup(b) {
        const stub = __rs(__dn(__f2p(import.meta.url)), 'stubs', 'BallooningSettingsStore.ts');
        b.onResolve({ filter: /BallooningSettingsStore$/ }, () => ({ path: stub }));
    },
};


const SRC = 'DSRFQ.Web/Modules/Common/Widgets/BallooningWidget/BallooningNumbering.ts';

const fails = [];
const check = (name, ok, detail = '') => {
    console.log(`  ${ok ? 'PASS' : 'FAIL'}  ${name}${detail ? '  ' + detail : ''}`);
    if (!ok) fails.push(name);
};

const dir = mkdtempSync(join(tmpdir(), 'balloonnum-'));
const out = join(dir, 'mod.mjs');

await build({
    entryPoints: [SRC],
    outfile: out,
    bundle: true,
    plugins: [stubStore],
    format: 'esm',
    platform: 'neutral',
    logLevel: 'warning',
});

const M = await import(pathToFileURL(out).href);

console.log('1. formatting and parsing');
check('a plain balloon prints as its integer',
    M.formatBalloonNumber({ balloonNumber: 5 }) === '5');
check('a child prints with the default separator',
    M.formatBalloonNumber({ balloonNumber: 5, subNumber: 1 }) === '5-1');
check('the separator is configurable',
    M.formatBalloonNumber({ balloonNumber: 5, subNumber: 1 }, '.') === '5.1');
check('subNumber 0 is not a child',
    M.formatBalloonNumber({ balloonNumber: 5, subNumber: 0 }) === '5');

const rt = (s) => M.formatBalloonNumber(M.parseBalloonNumber(s));
check('"5" round-trips', rt('5') === '5');
check('"5-1" round-trips', rt('5-1') === '5-1');
check('"5.1" is accepted even under the "-" setting', rt('5.1') === '5-1',
    `got ${rt('5.1')}`);
check('a plain number parses', M.parseBalloonNumber(7).balloonNumber === 7);
check('junk degrades to 0 rather than NaN',
    M.parseBalloonNumber('abc').balloonNumber === 0);
check('null degrades to 0', M.parseBalloonNumber(null).balloonNumber === 0);

console.log('\n2. ordering puts a child straight after its parent');
const nums = [
    { balloonNumber: 6 }, { balloonNumber: 5, subNumber: 2 },
    { balloonNumber: 5 }, { balloonNumber: 5, subNumber: 1 },
];
const ordered = [...nums].sort(M.compareBalloonNumber)
    .map(n => M.formatBalloonNumber(n));
check('5, 5-1, 5-2, 6', JSON.stringify(ordered) === '["5","5-1","5-2","6"]',
    ordered.join(', '));
check('nextSubNumber follows the highest child',
    M.nextSubNumber(nums, 5) === 3, String(M.nextSubNumber(nums, 5)));
check('nextSubNumber starts at 1 for a childless parent',
    M.nextSubNumber(nums, 6) === 1);

console.log('\n2b. a quantity expands into one line per feature');
const q = (n, qty) => ({ balloonNumber: n, quantity: qty });
const labels = (list) => M.expandByQuantity(list).map(x => x.label).join(' ');

check('no multiplier stays a single line',
    labels([q(44, undefined)]) === '44', labels([q(44, undefined)]));
check('quantity 1 stays a single line', labels([q(44, 1)]) === '44');
check('quantity 2 lists 44_1 and 44_2',
    labels([q(44, 2)]) === '44_1 44_2', labels([q(44, 2)]));
check('quantity 4 lists all four',
    labels([q(7, 4)]) === '7_1 7_2 7_3 7_4', labels([q(7, 4)]));
check('mixed set keeps balloon order',
    labels([q(1, undefined), q(2, 2), q(3, undefined)]) === '1 2_1 2_2 3',
    labels([q(1, undefined), q(2, 2), q(3, undefined)]));
// An instance and a sub-number must not look alike: 44-1 is an extra
// characteristic, 44_1 is the same one measured on another feature.
check('a child balloon expands under its own composite number',
    labels([{ balloonNumber: 5, subNumber: 1, quantity: 2 }]) === '5-1_1 5-1_2',
    labels([{ balloonNumber: 5, subNumber: 1, quantity: 2 }]));
check('the instance separator differs from the sub-number one',
    M.INSTANCE_SEPARATOR !== M.DEFAULT_SUB_SEPARATOR,
    `${M.INSTANCE_SEPARATOR} vs ${M.DEFAULT_SUB_SEPARATOR}`);

const ex = M.expandByQuantity([q(44, 2)]);
check('each line knows its position and total',
    ex[1].instance === 2 && ex[1].count === 2);
check('every line points back to the one balloon',
    ex[0].ann === ex[1].ann);
check('instanceCount reads the multiplier',
    M.instanceCount({ quantity: 3 }) === 3 && M.instanceCount({}) === 1);
check('a nonsense quantity counts as one',
    M.instanceCount({ quantity: 0 }) === 1
    && M.instanceCount({ quantity: -2 }) === 1
    && M.instanceCount({ quantity: NaN }) === 1);

console.log('\n3. grid cells parse in either order');
check('"D8" is letter-first',
    JSON.stringify(M.parseGridCell('D8')) ===
    JSON.stringify({ letter: 'D', number: 8, numberFirst: false }));
check('"8D" is number-first', M.parseGridCell('8D')?.numberFirst === true);
check('lowercase is normalised', M.parseGridCell('d8')?.letter === 'D');
check('UNMATCHED is not a cell', M.parseGridCell('UNMATCHED') === null);
check('blank is not a cell', M.parseGridCell('') === null);
check('formatGridCell inverts parseGridCell',
    M.formatGridCell(M.parseGridCell('8D')) === '8D');
check('letterOrdinal: A=1, Z=26, AA=27',
    M.letterOrdinal('A') === 1 && M.letterOrdinal('Z') === 26 &&
    M.letterOrdinal('AA') === 27);

console.log('\n4. grid system agrees with the Python GridSorter');
const g = M.parseGridSystem('F2', 'A1');
check('F2->A1 parses', !!g);
check('both axes descend, as the server reports',
    g.letterDirection === -1 && g.numberDirection === -1,
    JSON.stringify(g));
check('an unparseable pair yields null',
    M.parseGridSystem('', 'A1') === null);

const mk = (id, section, y, x, viewId = 0) => ({
    id, section, viewId, rect: { x, y, width: 1, height: 1 },
});
// The Python returned (0, 2, -6, 5, 5) for D8 under F2->A1.
const key = M.gridSortKey(mk('d8', 'D8', 5, 5), g);
check('D8 gives the same key the server computes',
    JSON.stringify(key) === JSON.stringify([0, 2, -6, 5, 5]),
    JSON.stringify(key));

console.log('\n5. unmatched balloons sort last but stay stable');
const set = [
    mk('u1', 'UNMATCHED', 1, 1),
    mk('p1', 'D8', 5, 5),
    mk('u2', '', 9, 9),
    mk('p2', 'B4', 2, 2),
];
const order = M.sortByGrid(set, g).map(a => a.id);
check('placed balloons come first',
    order.slice(0, 2).every(id => id.startsWith('p')), order.join(', '));
check('unmatched land at the end',
    order.slice(2).every(id => id.startsWith('u')), order.join(', '));
check('unmatched keep positional order among themselves',
    order.indexOf('u1') < order.indexOf('u2'), order.join(', '));
check('isUnmatched agrees',
    M.isUnmatched({ section: '' }) && M.isUnmatched({ section: 'UNMATCHED' }) &&
    !M.isUnmatched({ section: 'D8' }));

console.log('\n6. sorting is coherent when the grid is unknown');
const noGrid = M.sortByGrid(set, null).map(a => a.id);
check('no grid: everything falls back to position, nothing is dropped',
    noGrid.length === 4, noGrid.join(', '));

rmSync(dir, { recursive: true, force: true });

console.log(`\n${fails.length ? `${fails.length} FAILED: ${fails.join(', ')}` : 'ALL PASS'}`);
process.exit(fails.length ? 1 : 0);
